### ISSUE TITLE
ValueError: Failed to enter/exit configuration mode - Interface Range!

### ISSUE THREAD

**Issue by pourmirza** [ktbyers/netmiko#993]
Hi @ktbyers,

I am going to enable 802.1x authentication on Cisco 2960s switches. The Netmiko worked fine until I decided to use interface range command. I thought it makes it quicker if I use the 'interface range'. However,  I either get fail to enter or to exit configuration mode or sometimes the output isn't complete. I have enabled logging as you can see below. Please note I don't encounter any issue when I push the global configuration scripts and the error happens when using 'net_connect.send_config_set' to configure the interfaces using the' interface range'.

Code:

```
access_interface_config_commands_set = ['CDP enable\n',
'lldp receive\n',
'no switchport port-security\n',
'no switchport port-security maximum 4\n',
'no switchport port-security violation restrict\n',
'no switchport port-security aging time 5\n',
'no switchport port-security aging type inactivity\n',
'authentication control-direction in\n',
'authentication host-mode multi-auth\n',
'authentication event fail action next-method\n',
'authentication event server alive action reinitialize\n',
'authentication periodic\n',
'authentication timer reauthenticate server\n',
'authentication timer inactivity server dynamic\n',
'authentication order dot1x mab\n',
'authentication priority dot1x mab\n',
'authentication port-control auto\n',
'mab\n',
'dot1x pae authenticator\n',
'dot1x timeout tx-period 10\n',
'spanning-tree portfast\n',
'spanning-tree bpduguard enable\n']      
arranged_interfaces = {'int r Gi1/0/1-48': '2000'}
switch_config_number = 1
access_interface_config_commands_set_all = []
print('Switch {} Interface Configuration Started!'. format(switch_config_number), '\n')
for ar_int in arranged_interfaces:
    if 'Gi{}'.format(switch_config_number) in ar_int:
        access_interface_config_commands_set.insert(0, '{} \n'.format(ar_int))
        access_interface_config_commands_set.append('authentication event server dead action reinitialize vlan {} \n'.format(arranged_interfaces[ar_int]))
        access_interface_config_commands_set_all.extend(access_interface_config_commands_set)
        access_interface_config_commands_set.pop(0)
        access_interface_config_commands_set.pop(-1)
interface_conf_output = net_connect.send_config_set(access_interface_config_commands_set_all)
print(interface_conf_output, '\n')       
print('Switch {} Interface Configuration Completed!'. format(switch_config_number), '\n')   
```

Error:
```
--> 365         interface_conf_output = net_connect.send_config_set(access_interface_config_commands_set_all)
    366         print(interface_conf_output, '\n')
    367         print('Switch {} Interface Configuration Completed!'. format(switch_config_number), '\n')

c:\envs\nac\lib\site-packages\netmiko\base_connection.py in send_config_set(self, config_commands, exit_config_mode, delay_factor, max_loops, strip_prompt, strip_command, config_mode_command)
   1351         # Send config commands
   1352         cfg_mode_args = (config_mode_command,) if config_mode_command else tuple()
-> 1353         output = self.config_mode(*cfg_mode_args)
   1354         for cmd in config_commands:
   1355             self.write_channel(self.normalize_cmd(cmd))

c:\envs\nac\lib\site-packages\netmiko\cisco_base_connection.py in config_mode(self, config_command, pattern)
     40             pattern = re.escape(self.base_prompt[:16])
     41         return super(CiscoBaseConnection, self).config_mode(config_command=config_command,
---> 42                                                             pattern=pattern)
     43
     44     def exit_config_mode(self, exit_config='end', pattern='#'):

c:\envs\nac\lib\site-packages\netmiko\base_connection.py in config_mode(self, config_command, pattern)
   1269             output = self.read_until_pattern(pattern=pattern)
   1270             if not self.check_config_mode():
-> 1271                 raise ValueError("Failed to enter configuration mode.")
   1272         return output
  
logging output:
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:paramiko.transport:EOF in transport thread
````
incomplete output:

Switch 1 Interface Configuration Started!

config term
Enter configuration commands, one per line.  End with CNTL/Z.
Cisco2960s(config)#int r Gi1/0/1-48
Cisco2960s(config-if-range)#CDP enable
Cisco2960s(config-if-range)#lldp receive
Cisco2960s(config-if-range)#no switchport port-security
Cisco2960s(config-if-range)#no switchport port-security maximum 4
Cisco2960s(config-if-range)#no switchport port-security violation restrict
Cisco2960s(config-if-range)#no switchport port-security aging time 5
Cisco2960s(config-if-range)#no switchport port-security aging type inactivity
Cisco2960s(config-if-range)#authentication control-direction in

**Comment by ktbyers** [comment]
The exception you get is not consistent with the output you show from the session log?

Does your code only have a single thread? Do you have any sort of concurrency going on (threads/processes). Can you post your entire code perhaps in a simplified example that still exhibits the failure?

**Comment by pourmirza** [comment]
Hi @ktbyers ,

Thanks for your reply.

My code has only a single thread. I will post a simplified example in a few minutes. I managed to fix the issue by adding 'exit' at the end of 802.1x commands set.  I will test it on a few switches and let you know if I see the issue again.

```
access_interface_config_commands_set = ['CDP enable\n',
 'lldp receive\n',
 'no switchport port-security\n',
 'no switchport port-security maximum 4\n',
 'no switchport port-security violation restrict\n',
 'no switchport port-security aging time 5\n',
 'no switchport port-security aging type inactivity\n',
 'authentication control-direction in\n',
 'authentication host-mode multi-auth\n',
 'authentication event fail action next-method\n',
 'authentication event server alive action reinitialize\n',
 'authentication periodic\n',
 'authentication timer reauthenticate server\n',
 'authentication timer inactivity server dynamic\n',
 'authentication open\n',
 'authentication order dot1x mab\n',
 'authentication priority dot1x mab\n',
 'authentication port-control auto\n',
 'mab\n',
 'dot1x pae authenticator\n',
 'dot1x timeout tx-period 10\n',
 'spanning-tree portfast\n',
 'spanning-tree bpduguard enable\n',
 'exit\n']
```
Also, I have added delay_factor =4 and exit_config_mode =False into the config_set as below:

`interface_conf_output = net_connect.send_config_set(access_interface_config_commands_set_all, exit_config_mode=False, delay_factor=4)`

However, I have to send 'do wr' in order to save the config as the  exit_config_mode is set to False.

`net_connect.send_command('do wr')`

**Comment by ktbyers** [comment]
What version of Netmiko are you using?

So does it work when you set exit_config_mode=False? i.e. you no longer get the exception?

**Comment by pourmirza** [comment]
2.2.2
It works when I set exit_config_mode = False and add the 'exit' to my commands set scripts.  It is really weird as the  'net_connect.send_config_set' enter/exit the config mode automatically. However, I had to set the exit_config_mode to False and add the 'exit' to my script to fix this issue. It only happens when I use interface range command! I didn't see any issue when I used interface by interface approach.
 
I will test it on more switches and let you know the result.

**Comment by ktbyers** [comment]
Any update on posting the stack trace? I didn't see anything that jumped out at me as an underlying cause when I reviewed this earlier.

**Comment by pourmirza** [comment]
Hi @ktbyers ,

Sorry for my belated reply!  I had a really busy week.

I have tested the interface range command with no exit in the config set and without specifying exit_config_mode=False on a few switches. As you said, I kept the delay_factor=4 only as below:

`net_connect.send_config_set(config_commands, delay_factor=4)`

I can confirm that the scripts work fine with no errors. I guess it was the switch causing the issue.

Many thanks for your help.

Cheers,

Mo
